Improved Metering

AdamSLevy opened the pull request:
# Description

This PR refactors the Metering middleware to address the issues described in #999. In short, the current metering does not prevent execution of code segments that exceed the limit, but instead only checks if limits are exceeded after execution of the code segment. Additionally the current metering does not catch some situations where the limit exceeded in the last few instructions of a function.

This metering injects limit checks prior to each code segment, ensure that execution does not proceed if there is not enough gas to reach the next branching instruction. Although after an ExecutionLimitExceededError the reported gas used will be greater than the limit, the execution will not have actually proceeded through that code segment. 

Additionally, this metering will check further ahead than eWasm metering in certain circumstances.

For example if we have multiple nested loops in a block without any branching instructions,
```
METER
Block
    ... // no branching ops
    Loop
        METER
        ... // no branching ops
        Loop
            METER
            ...
            IF // first branching op since first METER
                ...
            END
        End
    End
End
```

Each METER represents injected metering code that adds the cost of the next segment to the used points and checks it against the limit, trapping if exceeded.

Each METER that I listed will actually ensure that there is enough gas to reach the first branching op, which in my above example is IF. 

I'm happy to answer more questions about the Metering design. I believe it would benefit from additional tests that use more complex wasm code.

## Implementation details

In order to implement this metering as such, a refactor of `parse::read_module()` and `codegen::MiddlewareChain.run()` and `codegen::EventSink` was required. 

This metering algorithm requires the ability to modify previously seen Events that have already been pushed onto the EventSink inside of feed_event. This means that this middleware chain needs to see all of the Events within a function prior to any subsequent chains seeing any of the Events. Additionally this means that all ParserState(Operator) events for a given function need to all be loaded into memory at the same time. 

In parse::read_module, I clone each Operator and store them in an events Vec<Event>. I load a complete function and then I pass the whole events Ven to MiddlewareChain.run(), which passes all events through each middleware chain. 

Unfortunately wasmparser::Operator does not implement clone, so I had to fork it and add the trait. I have opened an [issue on the latest wasmparser](https://github.com/bytecodealliance/wasmparser/issues/149) but for compatability I applied my patch to the [version that is currently](https://github.com/AdamSLevy/wasmparser/tree/OperatorClonev0.39.3) being used. 


## Issues

Currently my refactor of `parse::read_module` and `codegen::MiddlewareChain.run` breaks the singlepass backend. This is not due to the metering code but due to my refactor of the runtime-core, as I have isolated that patch in testing. Alternatively you can just comment out the line that adds the metering middleware entirely from the middleware-common-tests to show that the issue lies deeper. 

When running the middleware-common-tests compilation fails with singlepass I get the following failure:
```
Errors were encountered when committing before finalization: UnknownLabel(Dynamic(DynamicLabel(3)))
```
https://gist.github.com/AdamSLevy/455d2a50ec1211d20efaebcd9c06eb1c#file-gistfile1-txt-L42-L102

I have compared my code very closely with the master branch and I am 99% certain that all mcg and fcg functions are being called in the proper order in my refactor. The only difference I can see is that the Events are now all WasmOwned. 

However my metering code works perfectly with the LLVM backend. I have been working with @MarkMcCaskey to come up with this refactor and I have run into this issue and can't work past it as I believe it would require me to learn the singlepass backend and the dynasmrt Assembler. 

# Review

- [ ] Add a short description of the the change to the CHANGELOG.md file

AdamSLevy commented:
@ethanfrey this might interest you if you are using LLVM

losfair commented on lib/runtime-core/src/codegen.rs:
```diff
@@ -298,18 +298,14 @@ impl MiddlewareChain {
-        sink.push(ev);
         for m in &mut self.chain {
-            let prev: SmallVec<[Event; 2]> = sink.buffer.drain().collect();
+            let prev: Vec<Event> = sink.buffer.drain(..).collect();
```
Why is `SmallVec` replaced by `Vec` here? In most cases where no middleware made changes to the opcode stream, there should only be one Event here and an heap allocation would be saved by SmallVec.

losfair commented on lib/runtime-core/src/parse.rs:
```diff
@@ -266,28 +266,22 @@ pub fn read_module<
                     }
                 }
+
+                let events = EventSink { buffer: events };
```
Collecting all events into one big sink here would increase memory usage, especially for large functions. Is this really needed?

AdamSLevy commented:
@losfair I saw some comments you made on the code show up in my inbox but I don't see them here for some reason.

Thank you for taking the time for the review. 

Regarding use of Vec instead of SmallVec, this could be in error. I simply am new to Rust, and I knew that I'd be loading more than 2 Events into the Vec so that motivated the change.

Regarding the memory burden of loading all events of a function, I agree that this is not ideal. I think this could be reduced. 

The main goal here of this refactor is to allow middleware chains to be able to view some arbitrary number of opcodes to perform some static analysis, and then inject code at some other point. In the case of my Middleware, this means injecting metering code (gas used increase & limit check) with a placeholder value for the static cost, and then later replacing the placeholder value with the actual static cost once the end of the given code segment is reached. This is done by saving the index of the `const.i64` instruction that saves the cost, and then referring back to it to update it on some later call of `feed_event()`.

My current design relies on the caller of `feed_event` (the `codegen::MiddlewareChain.run` function), to allow EventSink to serve as a place that the chain can store Events without them being flushed. For simplicity of implementation I simply loaded all Events of a function.

I believe there is a better implementation which allows the chain to control how many Operators get loaded into memory, which would be less than an entire function in most cases. Instead of loading the entire function, a middleware can have its own Vec<Events> that it can clone Events into to cache them. It can then simply not push them onto the given EventSink until enough Events have been observed for the purposes of the middleware. In the case of this metering middleware, it would only need to cache as many Events as the longest non-branching code-segment. With this, my changes to runtime-core::codegen and runtime-core::parse could be reverted to the current implementation.

I actually originally attempted something like this but I ran into lifetime issues, and this was before I patched wasmparser to add the Clone trait to Operator.

MarkMcCaskey commented:
@losfair I think if we have the Opcodes as Clone, we can do the iterator style I prototyped before.  That should avoid having to store them all and shouldn't be any slower in the case of no middleware; I ran into unresolvable issues due to the lifetimes too when trying this out but it should work if they're Clone

AdamSLevy commented:
> Instead of loading the entire function, a middleware can have its own Vec that it can clone Events into to cache them. It can then simply not push them onto the given EventSink until enough Events have been observed for the purposes of the middleware.

I attempted to implement this but I could not get around lifetime issues.

Events can't seem to be stored in a Vec<Event<'a,'b>> that is a member of the Metering struct. Also putting them back on the EventSink creates lifetime conflicts in the same way. This is a frustrating limitation of storing Events in a Vec that is a member of the Metering struct. The Metering struct demands a lifetime that is unrelated to that of the incoming Events. This is why allowing them all to be put in the EventSink that MiddlewareChain.run provides gets around lifetime constraints of the Events.

MarkMcCaskey commented:
@AdamSLevy if Opcode from wasmparser is Clone then we can remove one of the lifetimes on `Event`.  We can only use the owned one because it's just way simpler.  I believe that should fix things!

AdamSLevy commented:
I'm not sure that will resolve all of the lifetime issues, but that would simplify things.

AdamSLevy commented:
If Event itself could also be cloned, that would make things easy.

MarkMcCaskey commented:
I agree, we should make Event Clone!  I can't remember a reason besides Opcode not being Clone that it wasn't

AdamSLevy commented:
I tried it and ran into an issue with InternalFunction::Breakpoint.

MarkMcCaskey commented:
@AdamSLevy we can fix that by changing 
```Rust
pub type BreakpointHandler =
    Box<dyn Fn(BreakpointInfo) -> Result<(), Box<dyn Any>> + Send + Sync + 'static>;
```

Into 
```Rust
use std::rc::Rc;
pub type BreakpointHandler =
    Rc<dyn Fn(BreakpointInfo) -> Result<(), Box<dyn Any>> + Send + Sync + 'static>;
```

I believe.  Box is a heap allocation so copying it doesn't make much sense, Rc is like a shared heap allocation!

I'm not sure what the impacts of a change like are.  If we need mutable access to the thing then we'll need `Rc<RefCell<...>>` and if we need to share it between threads then we'll need `Arc<Mutex<...>>`

ethanfrey commented:
I'm not sure why it doesn't work on singlepass and I definitely need to dig in there. I think the dynasm is the same reason it doesn't serialize.

I would be happy to test this out with llvm support, but given the inability to import llvm backend from crates.io (it cannot be published), I won't likely be using llvm in an app until that is resolved. But I can try to make a test case of some malicious wasm if that would help

AdamSLevy commented:
@ethanfrey Any contribution to figuring out why singlepass is not working or testing the robustness of this metering algorithm would be greatly appreciated. I have wanted to add more tests that establish that metering is actually preventing execution past certain points. 

For example, in the following pseudo-code, the metering injection algorithm should prevent execution past the first METER if there is not enough gas to reach the first branching statement (an `if` in the example), which may be nested within one or more loop.  
```
const.i32 0
block
    METER // ensures that there is enough gas to reach the next branching op
    ... // non-branching ops, possibly other loops and METERs that all ensure enough gas to reach the next branching op
    loop
        METER // ensures enough gas to reach the if
        ... // non-branching ops
        if // first branch
...
```
A good test could increment a variable between these METER statements and check that it was not incremented past a certain point based on various limits.

What is this issue regarding llvm? I am planning to use my forked version of wasmer that includes exposing this backend with metering through to Golang for my application. I haven't tested this on Windows yet, which I'd like to support. Will what you mentioned cause me any issues?

AdamSLevy commented:
@MarkMcCaskey If you publish a branch that adds the Clone trait to Event based on the latest `master`, I can attempt to refactor this metering to store Events internally in the Metering struct. This would reduce memory usage to whatever is required to store the Events of contiguously non-branching code sequences, as opposed to an entire function and it may happen to resolve the singlepass backend issue.

losfair commented:
@AdamSLevy I deleted my previous comment because I didn't fully read your changes by then. Sorry if there is any misunderstanding on that.

I think if `Event` is `Clone`, we won't really need to change the current `master`'s middleware system itself. The Metering middleware can simply cache everything within a local buffer until a terminator is found.

For the issue with singlepass failing to run cached code with metering injected, I think it is because the way singlepass handles `InternalEvent::Breakpoint` depends on runtime memory. This needs a fix and before that a workaround is to ensure `singlepass & (metering ^ caching)`.

AdamSLevy commented:
Aside from the issues discussed above, I have discovered that my metering algorithm has a bug that causes it to add too much gas to the points used. I'm working on a refactor for that issue.

AdamSLevy commented:
The latest force push rebases this PR on the latest master and also fixes the bugs in my metering algorithm that I mentioned in my last comment.

syrusakbary commented on Cargo.toml:
```diff
@@ -110,3 +110,6 @@ crate-type = ["bin"]
 crate-type = ["bin"]
+
+[patch.crates-io]
+wasmparser =  { git = "https://github.com/AdamSLevy/wasmparser", branch = "OperatorClonev0.39.3" }
```
Can you create a PR into `wasmparser` so we don't need to patch it here?

https://github.com/bytecodealliance/wasmparser/compare/master...AdamSLevy:OperatorClonev0.39.3

AdamSLevy replied:
https://github.com/bytecodealliance/wasmparser/pull/157

AdamSLevy replied:
The issue you may find with my PR is that it is based on the latest wasmparser version. This version needs to match the version given by some external dependencies as well. That locks us to 0.39.3 atm, hence my branch based on that version specifically.

AdamSLevy replied:
This is now merged upstream.

AdamSLevy commented:
The latest force push simply rebased on master and added a small fix that properly re-initializes the metering state at the start of each function.

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
 /// Each compiler backend with Metering enabled should produce the same cost used at runtime for
 /// the same function calls so we can say that the metering is deterministic.
-///
 pub struct Metering {
```
Please add `#[derive(Debug)]` (lets us print it out in debug contexts), generally we want this on all of our types because it applies recursively!

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
 impl Metering {
-    pub fn new(limit: u64) -> Metering {
+    pub fn new() -> Metering {
         Metering {
```
Small improvement, but you can use the keyword `Self` in `impl X` contexts to refer to the type being created, so this can be 

```Rust
pub fn new() -> Self {
    Self { ... }
}
```

One more improvement, is that we can add `#[derive(Default)]` to the struct (should look like `#[derive(Debug, Default)]`) and then do something like:

```Rust
pub fn new() -> Self {
   Self::default()
}
```

These are very minor points, but it makes editing the code a bit quicker by reducing redundancy; and if in the future we want a default struct but with some number of fields not default, we can use the syntax:

```Rust
Self {
   override_field1: value1,
   override_field2: value2,
   ..Self::default()
}
```

AdamSLevy replied:
Thanks! I really appreciate the suggestions so I can improve my Rust skills.

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
+        }
+    }
+
+    fn set_costs<'a, 'b: 'a>(&mut self) {
```
I don't believe these lifetime annotations, `<'a, 'b: 'a>` are used in this function, so we can remove them.

Because this is its own `impl` and not a trait method, we aren't required to make the type signature conform to anything extra

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
+                add: self.current_block_cost,
+                check: 0,
+            };
+        }
```
⛳️ 
```Rust
if let Some(last) = self.current_block_injections.last_mut() {
    *last = Injection {
           add: self.current_black_cost,
           check: 0,
     };
}
```

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
+        self.current_block_cost = 0;
+    }
+
+    fn begin<'a, 'b: 'a>(&mut self) {
```
Same here, I think we can remove these lifetime annotations

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
+        self.current_block_injections
+            .push(Injection { add: 0, check: 0 });
+    }
+    fn end<'a, 'b: 'a>(&mut self) {
```
Same here, I think we can remove these lifetime annotations

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
+                    sink.push(ev);
+                    self.injections[0].inject(sink);
+                }
+                Event::Wasm(&ref op) | Event::WasmOwned(ref op) => match *op {
```
⛳️ 
I believe `&ref op` is exactly the same as `op` I think, `ref` makes it a reference and `&` pattern matches the reference away

AdamSLevy replied:
I thought so too but I rant into issues when I used any different combination here. I think the compiler requires the `&ref op` on `Event::Wasm` to match `Event::WasmOwned(ref op)` but I can't precisely articulate why.

This exact match pattern was used in the original metering code.

In this particular branch I don't believe I will ever see any `Event::Wasm` as everything is `Wasm::Owned` once it is passed to the middleware in my current refactor of `codegen`, but I didn't want to assume that would always be the case in the middleware.

MarkMcCaskey replied:
Okay, that makes sense -- I find it easy to believe that when matching on multiple things to the same body that the types need to line up exactly and somehow that this makes it work 🤷‍♂ .  Thanks for looking into it!

MarkMcCaskey commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
+        let prev: Vec<Event> = sink.buffer.drain(..).collect();
+        let mut inj_idx: usize = 1;
+        for ev in prev {
+            match ev {
```
It looks like we could put `sink.push(ev);` here and remove it everywhere else.  Feel free to not do that if you think that there will be branches in the future where that's not the case

AdamSLevy replied:
In some cases the event still needs to be known after it is pushed, which would move the event and make it unusable. So I opted for being explicit and putting `sink.push(ev)` every place it is needed over trying to find a more clever way to reduce code.

MarkMcCaskey replied:
Oh that's fair -- if you need it in the branches, then you can probably do it at the end of the loop 🤔 ; yeah, I was just thinking that if we want to push it every time, it's more "obviously correct" if the push happens unconditionally somewhere.  As it is now, the reader has to read all the branches to find out what the logic is in regards to the event and when editing has to make sure to do it in every new branch.  If it makes sense for there to be branches where it will not be pushed, then I think it's okay, but otherwise I think having it in one place is easier to understand.

AdamSLevy replied:
I'll take another look and see if I can't find a more elegant solution.

MarkMcCaskey replied:
I just took another look and I see what you're saying now, if it doesn't make sense to do this, then that's fine too!

MarkMcCaskey commented on lib/runtime-core/src/codegen.rs:
```diff
@@ -274,7 +273,8 @@ fn requires_pre_validation(backend: Backend) -> bool {
 pub struct EventSink<'a, 'b> {
-    buffer: SmallVec<[Event<'a, 'b>; 2]>,
+    /// document this
+    pub buffer: Vec<Event<'a, 'b>>,
```
We can talk about this later, but it may make sense to keep using a SmallVec, but maybe a larger SmallVec depending on how everything works; I'm mostly reviewing for Rust style right now so I didn't catch the event sink behavior

AdamSLevy replied:
Sure. I am ambivalent about what vector is used as I am still mostly ignorant to the subtle differences.

MarkMcCaskey replied:
In this case it's a very small thing, SmallVec is just a combination of static array and vector, so that you can avoid heap allocations in most cases while still supporting outliers that need an arbitrarily large vector.

It's especially useful if you're writing a generic data type which under normal circumstances only needs to hold a little bit of data, but it's possible that it could sometimes hold a lot or that some other code could use it in ways you didn't anticipate.  I'm not sure useful it is here 🤷‍♂

MarkMcCaskey commented on lib/runtime-core/src/parse.rs:
```diff
@@ -266,28 +266,22 @@ pub fn read_module<
+                            events.push(Event::WasmOwned(op.clone()));
+                        }
+                        ParserState::EndFunctionBody => {
+                            events.push(Event::Internal(InternalEvent::FunctionEnd));
```
I think we can keep the old style where we call middlewares at each step to prevent filling a buffer with the wasm ops.  That way middleware can do its own lookahead.  The one structural change we'll need to make is ensuring that middleware and our code can handle being given nothing or given more at once.

I'm still fond of using iterators, something like https://github.com/wasmerio/wasmer/pull/1021 , but I ran into issues with all the traits the last time I touched that and due to wasm op not being clone, it wasn't possible

AdamSLevy replied:
>  That way middleware can do its own lookahead. 

I am still unsure how to achieve this with the current streaming design of WasmParser. 

> I'm still fond of using iterators, something like #1021 

How would this allow for a middleware to inject code? How would that affect the iterator? Perhaps this is a rudimentary Rust question, in which case just let me know and I'll do the appropriate Googling.

MarkMcCaskey replied:
The way middleware would work is by allowing any link in the chain to hold on to things.

So for example given the chain:

```
Parser -> A -> B -> C -> CodeGen
```

Any step can hold on to as many events as it likes.  So if `A` needs to have full lookahead then it would read from Parser until parser stopped returning anything and then start returning to `B`.

If `A` needs to look at functions and `B` needs to look at loops, then `A` can track the state of the streaming opcodes, when it sees a `function start` event it caches events until it sees `function end` and then dumps everything (including any modifications it made into B).  B does the same with loops and so on.

The Rust Iterator trait defines a function `next(&mut self) -> Option<Self::Item>` with the logic being that it returns `Some(T)` while it has items to return and returns `None` when there's nothing left to return and the iterator is done.

Using iterators, we'd have the `CodeGen` step get an opaque iterator and call `next` on it doing whatever it wants with it to generate code.

That iterator is step `C`, what `C` does is use its internal iterator to get events and then do logic on them, so when `C`.next() is called, `B`.next() is called which calls `A`.next() which calls `Parser`.next().  At any step in this chain, an iterator can call `next` on its internal iterator as many times as it wants allowing it to lookahead as much as it wants.  To be more specific about how it'll commonly work: once it's read all that it wants to into its internal buffer, the struct implementing it sets a flag or otherwise tracks that it's ready to return and returns the first item.  When that bit is set, it doesn't call `next()`, it just dumps its internal buffer/outputs its own edits and additions to the stream.

I like this API design a lot because it's very elegant, none of the links in the chain need to know about anything other than how to get more input.

MarkMcCaskey replied:
And because Iterator is well supported in Rust, if you wanted to just get everything all at once, that's very easy to do, you can just do `internal_iter.collect::<Vec<_>>()` and you have a vector of everything.  In the future when generators and `yield` are in stable Rust, this type of API can be even nicer.  Instead of making an iterator explicitly, you just write a normal function but use the keyword `yield` instead of `return` for each element and magic happens that tracks the state of execution in your function for you.

AdamSLevy replied:
Sounds like this is how it should work. I am not sure how to implement it but if you can then I will modify the middleware to use that approach.

MarkMcCaskey commented:
> @MarkMcCaskey If you publish a branch that adds the Clone trait to Event based on the latest `master`, I can attempt to refactor this metering to store Events internally in the Metering struct. This would reduce memory usage to whatever is required to store the Events of contiguously non-branching code sequences, as opposed to an entire function and it may happen to resolve the singlepass backend issue.

Sorry, I completely missed this comment!  I'll look into it

MarkMcCaskey commented:
bors try

MarkMcCaskey commented:
> > @MarkMcCaskey If you publish a branch that adds the Clone trait to Event based on the latest `master`, I can attempt to refactor this metering to store Events internally in the Metering struct. This would reduce memory usage to whatever is required to store the Events of contiguously non-branching code sequences, as opposed to an entire function and it may happen to resolve the singlepass backend issue.
> 
> Sorry, I completely missed this comment! I'll look into it

Following up on this, I made #1080 ; once that ships we can do the rest of the changes on master

AdamSLevy commented:
> > > @MarkMcCaskey If you publish a branch that adds the Clone trait to Event based on the latest `master`, I can attempt to refactor this metering to store Events internally in the Metering struct. This would reduce memory usage to whatever is required to store the Events of contiguously non-branching code sequences, as opposed to an entire function and it may happen to resolve the singlepass backend issue.
> > 
> > 
> > Sorry, I completely missed this comment! I'll look into it

Actually we later discussed this on Discord. I added the Clone trait to Event using your above suggested modification of BreakpointHandler. However I was not able to get the Metering struct to own events with lifetimes defined by the caller of feed_event across calls due to lifetime issues.

> Following up on this, I made #1080 ; once that ships we can do the rest of the changes on master

This is good as it gets us the Clone trait on wasmparser::Operator, but I still do not know how the parse and codegen modules could be refactored to allow middleware to own/cache these Events between calls to feed_event.

Thanks for the thorough review. I'll get the above suggested changes pushed and rebased on master soon.

webmaster128 commented:
I guess this PR is outedated now there is a new metering middleware in master. But is the essence of this work included in the new metering schema?

syrusakbary commented:
Yeah, this PR is outdated now as is based on the Wasmer 0.x codebase which doesn't include the new structure brought by the refactor for 1.0.

I believe the essence is in included in the new metering schema.
Here's the new code in case is useful for review: https://github.com/wasmerio/wasmer/blob/master/lib/middlewares/src/metering.rs#L182-L216

New PRs would be more than welcome for improving the current metering middleware.
Closing this PR now :)

AdamSLevy commented on lib/middleware-common/src/metering.rs:
```diff
@@ -1,37 +1,118 @@
+                }
+                Event::Wasm(&ref op) | Event::WasmOwned(ref op) => match *op {
+                    Operator::End
+                    | Operator::If { .. }
```
@syrusakbary 

It's been awhile since I had all the mental models loaded in my head, and I don't have the context on the new metering code you linked. But I believe that If needs to be included in the following list or else I think it may suffer a similar issue that this code was attempting to address, or at minimum calculate the code cost incorrectly. Of course the implications of adding that If operator would need to be considered in the context of the rest of the metering code. 

The reason I think it needs to be there is that an If technically starts a new section/branch of code. If the condition for the if is false, it will result in a jump. 

https://github.com/wasmerio/wasmer/blob/91ffef92efaa3803e541268c33a7b2af07d77eab/lib/middlewares/src/metering.rs#L182-L216
